### ISSUE TITLE
GAN training is a little unstable.

### ISSUE THREAD

**Issue by ryanwebster90** [boschresearch/unetgan#1]
Hey there,
Awesome work, the results on FFHQ 256x256 are stunning. I was able to get this code to converge on FFHQ to a similar quality as shown in the paper (after 3 or so random seeds which diverged). On the other hand, I have not had luck with other datasets, such as LSUN tower/cat. I tried "warm-starting" the GAN with the FFHQ pre-trained (a trick that normally works pretty nice), which at least generated recognizable images, but would not improve after about 50k iterations. 
I'd love to use this method, as a competitor/alternative to stylegan2, but I'm afraid it is too fickle at the moment. Any tricks I'm missing? 

My docker image is pytorch/pytorch latest and I use ffhq training script in training_scripts. I'm using single gpu training on a 1080 Ti (with 11GB gpu RAM), so I've set the batchsize to 4. Perhaps this is the source of instability.

**Comment by edgarschnfld** [comment]
Hi, 
Yes, training on FFHQ is not as stable as let's say CelebA, and the batch size does play a role. This phenomenon already exists with the original BigGAN, on which U-Net GAN builds. The FID curves in the readme show what a good and bad curve looks like. If a curve looks like the good curve initially but has a major hickup around 50K, don't kill the experiment because the FID will most likely recover to follow it's normal path again. Maybe this is what you are observing?. Depending on the dataset, using ```--consistency_loss``` instead of ```--consistency_loss_and_augmentation``` can be more stable, and you should give it a try. In this case, also remove ```--full_batch_mixup```. Alternatively, if one were to keep the augmentation aspect, then setting the ```--slow_mixup_epochs``` to a higher number may be helpful. This is because the 2D cutmix augmentation can be too much to handle at the early training training steps.
A batch size of 4 is indeed very low. My observation is that the batch size does affect stability. I suspect that this is the reason that the original BigGAN used a batch size of 2048: It means with 1000 classes, the conditional batch norm sees on average 2 images per class. That's why COCO-Animals works with a way smaller normal batch size - there are only 10 classes - but with too small of a batch size it also becomes unstable. For FFHQ, the batch norm is unconditonal (well technically we condition it on noise, aka self-normalization (see https://arxiv.org/abs/1810.01365)), which is why training with small batch sizes works, but 4 seems too small still. A way to test this would be to reduce the resolution to 128 in order to increase the batch size and see if you observe better results. 
Best of luck :)

**Comment by ryanwebster90** [comment]
Hey, 
thanks for the response. I'll re-try a few configurations with your recommendations on the more challenging datasets.
